### ISSUE TITLE
UCX PML add_procs bug

### ISSUE THREAD

**Issue by bwbarrett** [open-mpi/ompi#7570]
While digging into an odd add_procs behavior for the ofi mtl and the cm pml, I ran across what I'm pretty sure is a bug (or two) in the UCX PML on master.  The issue is in mca_pml_ucx_add_proc() on [line 403](https://github.com/open-mpi/ompi/blob/c704ed4cc51baf3ee0926454476b2bee0148581f/ompi/mca/pml/ucx/pml_ucx.c#L403) where the UCX PML is calling pml_check_selected().  First issue is that pml_check_selected() expects the proc being added, not proc 0, to be the one passed into the function.  The second issue is that the third argument is supposed to be the number of entries in the procs array, not the remote rank (the comment about not using the third argument isn't guaranteed, but is true today).

The impact of the first issue is that you've essentially disabled the PML is the same check when using late-binding proc creation today, and will cause a return of the rank 0 direct call flood issue in the near future.  Today, the code got better because George added a temporary hack to the check_selected call to return success if the key wasn't already locally loaded.  Soon, we are finishing the fix and replacing that with a blocking call to the added rank (which has already been filled in and therefore is a quick call).  However, this change will cause a return of the "everyone pounds on rank 0" problem for the UCX PML, as it is passing the wrong data to check_selected().  Worse, because the UCX PML is passing an erroneous procs array length, the fixes mentioned above are likely to start causing segfaults.

I have no way of testing changes to the UCX PML; can someone who does please take a look at this issue?

**Comment by bosilca** [comment]
I have a patch, but right now I am unable to run any OMPI based code so I can't validate it. I also allowed compilations with newer versions of UCX, I've been using them (including UCX master) successfully for quite some time.

```diff
diff --git a/ompi/mca/pml/ucx/pml_ucx.c b/ompi/mca/pml/ucx/pml_ucx.c
index 93a2760..fb97299 100644
--- a/ompi/mca/pml/ucx/pml_ucx.c
+++ b/ompi/mca/pml/ucx/pml_ucx.c
@@ -393,22 +393,6 @@ static ucp_ep_h mca_pml_ucx_add_proc_common(ompi_proc_t *proc)
     return ep;
 }

-static ucp_ep_h mca_pml_ucx_add_proc(ompi_communicator_t *comm, int dst)
-{
-    ompi_proc_t *proc0     = ompi_comm_peer_lookup(comm, 0);
-    ompi_proc_t *proc_peer = ompi_comm_peer_lookup(comm, dst);
-    int ret;
-
-    /* Note, mca_pml_base_pml_check_selected, doesn't use 3rd argument */
-    if (OMPI_SUCCESS != (ret = mca_pml_base_pml_check_selected("ucx",
-                                                               &proc0,
-                                                               dst))) {
-        return NULL;
-    }
-
-    return mca_pml_ucx_add_proc_common(proc_peer);
-}
-
 int mca_pml_ucx_add_procs(struct ompi_proc_t **procs, size_t nprocs)
 {
     ompi_proc_t *proc;
@@ -438,24 +422,26 @@ static inline ucp_ep_h mca_pml_ucx_get_ep(ompi_communicator_t *comm, int rank)
 {
     ucp_ep_h ep;

-    ep = ompi_comm_peer_lookup(comm, rank)->proc_endpoints[OMPI_PROC_ENDPOINT_TAG_PML];
-    if (OPAL_LIKELY(ep != NULL)) {
-        return ep;
+    if (OPAL_UNLIKELY(rank >= ompi_comm_size(comm))) {
+        PML_UCX_ERROR("Rank number (%d) is larger than communicator size (%d)",
+                      rank, ompi_comm_size(comm));
+        return NULL;
     }

-    ep = mca_pml_ucx_add_proc(comm, rank);
-    if (OPAL_LIKELY(ep != NULL)) {
+    ompi_proc_t *proc_peer = ompi_comm_peer_lookup(comm, rank);
+    ep = proc_peer->proc_endpoints[OMPI_PROC_ENDPOINT_TAG_PML];
+    if (OPAL_LIKELY(NULL != ep)) {
         return ep;
     }

-    if (rank >= ompi_comm_size(comm)) {
-        PML_UCX_ERROR("Rank number (%d) is larger than communicator size (%d)",
-                      rank, ompi_comm_size(comm));
-    } else {
-        PML_UCX_ERROR("Failed to resolve UCX endpoint for rank %d", rank);
+    /* Note, mca_pml_base_pml_check_selected, doesn't use 3rd argument */
+    if (OMPI_SUCCESS != mca_pml_base_pml_check_selected("ucx",
+                                                        &proc_peer,
+                                                        1)) {
+        return NULL;
     }

-    return NULL;
+    return mca_pml_ucx_add_proc_common(proc_peer);
 }

 int mca_pml_ucx_del_procs(struct ompi_proc_t **procs, size_t nprocs)
diff --git a/opal/mca/btl/uct/configure.m4 b/opal/mca/btl/uct/configure.m4
index 61c7c8c..f125f3b 100644
--- a/opal/mca/btl/uct/configure.m4
+++ b/opal/mca/btl/uct/configure.m4
@@ -52,7 +52,7 @@ dnl
     fi

     max_allowed_uct_major=1
-    max_allowed_uct_minor=7
+    max_allowed_uct_minor=8
     if test "$btl_uct_happy" = "yes" && test "$enable_uct_version_check" != "no"; then
         AC_MSG_CHECKING([UCT version compatibility])
         OPAL_VAR_SCOPE_PUSH([CPPFLAGS_save])
```

**Comment by yosefe** [comment]
@bosilca added fix #7575, i've removed the ` if (OPAL_UNLIKELY(rank >= ompi_comm_size(comm))) {' check since it's in ompi_comm_peer_lookup() anyway in debug mode, and in release mode we'd like to avoid the extra check.
